[novafinedust] Nova Fine Dust binding for SDS011 sensors

t2000 opened the pull request:
Closes #7527

Signed-off-by: Stefan Triller <github@stefantriller.de>

t2000 commented:
I have the binding attached for testing, please rename to ".jar" before putting it into the addons folder.

[org.openhab.binding.novafinedust-2.5.5-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4567974/org.openhab.binding.novafinedust-2.5.5-SNAPSHOT.zip)

fwolter commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,287 @@
+                serialPort = null;
+            }
+        }
+        IOUtils.closeQuietly(inputStream);
+        IOUtils.closeQuietly(outputStream);
```
Since we are going to remove the Apache Commons dependencies, could you replace the lines with standard Java code? See #7722

t2000 commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,244 @@
+                    try {
+                        communicator.requestSensorData();
+                    } catch (IOException e) {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
```
it will be set to ONLINE again once there is valid data received, see method `updateChannels`.

t2000 commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,244 @@
+        }
+    }
+
+    private boolean validateConfiguration() {
+        if (config.port == null || config.port.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, "Port must be set!");
+            return false;
+        }
+
+        if (config.reporting) {
+            if (config.reportingInterval < 0 || config.reportingInterval > 30) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                        "Reporting interval has to be between 0 and 30 minutes");
+                return false;
+            }
+        } else {
+            if (config.pollingInterval < 3 || config.pollingInterval > 3600) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                        "Polling interval has to be between 3 and 3600 seconds");
+                return false;
+            }
+        }
+        return true;
+    }
```
Oh wow, so openHAB now validates these values before passing them to the handler? That's cool, I didn't know that this was implemented. Thanks for letting me know. I have removed the checks.

fwolter replied:
I'm afraid I have to retract my words. The config is only validated in Paper UI, but not when configuring via a things file. @cpmeister should bindings double check their config?

cpmeister replied:
It is always better if the binding double checks the configs.

t2000 replied:
Alright, I have reverted the removal of the check, so it is back in place.

t2000 commented:
Thanks for the review. I have addressed your comments, except for one where I replied here on github.

I have also rebased the code against the latest 2.5.x branch.

fwolter commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -253,7 +252,7 @@ public void serialEvent(SerialPortEvent event) {
                 logger.debug("Got data from sensor: {}", reply);
             } catch (IOException e) {
-                logger.error("Could not read available data from the serial port", e);
+                logger.warn("Could not read available data from the serial port", e);
```
No need to log the stacktrace here, but you could log the exception message:
```suggestion
                logger.warn("Could not read available data from the serial port: {}", e.getMessage());
```

t2000 replied:
Indeed, this makes sense to only log the message. I have changed it, thanks!

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,296 @@
+            logger.warn("Problem while waiting for reading a reply to our command.");
+            Thread.currentThread().interrupt();
+        }
+        SensorReply reply = readReply();
```
Wouldn't this already be called implicitly through `serialEvent`? Wouldn't the data on the input stream be gone at that point? Could you explain how this works? I might be misunderstanding how the serial protocol works.

t2000 replied:
I am configuring the sensor with the commands BEFORE I register the listener.

I found it a lot easier to do the protocol stuff in a synchronous manner so I have commands and their replies in one place to check if they belong together and there are no errors.

So this is all done before the calls:
```
localSerialPort.notifyOnDataAvailable(true);
localSerialPort.addEventListener(this);
```

cpmeister replied:
Ah! Didn't notice that, makes a lot more sense now, thanks!

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+				<label>Reporting if true, otherwise polling</label>
+				<description>Reporting is strongly recommended to increase sensor lifetime</description>
+			</parameter>
+			<parameter name="reportingInterval" type="integer" min="0" max="30">
```
```suggestion
			<parameter name="reportingInterval" type="integer" min="0" max="30" unit="min">
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+				<label>Time in minutes when a value will be reported</label>
+				<description>Device will report every x minutes and sleep for x*60 - 30 seconds afterwards, 0 = as fast as possible without sleep</description>
+			</parameter>
+			<parameter name="pollingInterval" type="integer" min="3" max="3600">
```
```suggestion
			<parameter name="pollingInterval" type="integer" min="3" max="3600" unit="s">
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+			<parameter name="reportingInterval" type="integer" min="0" max="30">
+				<default>1</default>
+				<advanced>true</advanced>
+				<label>Time in minutes when a value will be reported</label>
```
The first letter of every word in a label should be capitalized.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+			</parameter>
+			<parameter name="reporting" type="boolean">
+				<default>true</default>
+				<label>Reporting if true, otherwise polling</label>
```
The first letter of every word in a label should be capitalized.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+			<parameter name="pollingInterval" type="integer" min="3" max="3600">
+				<default>10</default>
+				<advanced>true</advanced>
+				<label>Time in seconds when a value will be polled from the device (min 3 seconds)</label>
```
The first letter of every word in a label should be capitalized.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/WorkMode.java:
```diff
@@ -0,0 +1,24 @@
+ * @author Stefan Triller - Initial contribution
+ *
+ */
+public enum WorkMode {
```
Please add `@NonNullByDefault`

t2000 commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,296 @@
+        outputStream.flush();
+        try {
+            // Give the sensor some time to handle the command
+            Thread.sleep(500);
```
Unfortunately this doesn't work. It is probably a "feature" of the sensor. So I have to wait here before reading the reply, otherwise the communication with the sensor will be broken.

cpmeister replied:
I'm still confused, how does reading data from a serial port break communication with the sensor. Are you sure this isn't a bug in the gnu.io serial library?

t2000 commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,228 @@
+        }
+
+        if (communicator != null) {
+            scheduler.schedule(() -> communicator.dispose(), 0, TimeUnit.SECONDS);
```
Because I think `initialize` and `dispose` from a `ThingHandler` should return quickly, i.e. within 5 seconds. Inside the `dispose` I send the sensor to sleep by communicating with it. This is important to enhance its lifetime.

So in order to not block the disposing call, I do it asynchronously.

cpmeister replied:
So what happens if the handler is restarted (disposed then initialized again)? Would the handler be stuck in a configuration COMMUNICATION_ERROR because the port is still in use?

t2000 replied:
This is a VERY interesting point! :-)

There are similar problems in the zigbee binding, because there it also takes a while to startup the stack and if this gets restarted through dispose/initialize calls, everything gets messed up too. Synchronization might help there but is a pain in the ass to do that for every binding.

And I am afraid to say that with the 5 second rule for initialize/dispose, bindings are forced to spawn a separate task, which will ALWAYS cause these troubles if dispose/initialize is called within a short time frame (i.e. by a disable/enable of the thing).

If you prefer to have my dispose here running in sync with the framework, I am fine to replace the task with a direct call. In my case this usually only takes about 1-2 seconds anyway, but I wanted to play safe since I am communicating with an external device and this might timeout, etc.

But the overall problem for all bindings will stay, especially with the use case of enabling and disabling of a thing.

WDYT?

cpmeister replied:
The only way I can see of dealing with delayed initialize/dispose is to make your handler and communicator objects have different lifecycles.

You would need to create a your own communicator factory class that handlers would request communicators for a given port. 
```
public interface CommunicatorFactory{
     SDS011Communicator getCommunicator(SerialPortIdentifier portId);
}
```

The factory itself would share the same lifecycle as an osgi component so either you could make your handler factory a communicator factory or implement it in a new osgi component.

The key benefit here is that you would be able to reuse SDS011Communicator instances across handler restarts.

t2000 replied:
This doesn't really solve the problem, I could still have the communicator instance running if i dispose the handler and then the port is blocked.

Also its not worth the effort as I said before the communication takes max 1-2 seconds.

I have changed it to be synchronous now.

cpmeister replied:
>This doesn't really solve the problem, I could still have the communicator instance running if i dispose the handler and then the port is blocked.

Not really since you can prevent a communicator from opening the port if it is already open. Likewise a communicator can stop its disposal process if a new handler requests to use it.

>I have changed it to be synchronous now.

I don't think would work, a new handler means a new communicator object. Synchronization only works on the same object instance.

t2000 replied:
> Not really since you can prevent a communicator from opening the port if it is already open. Likewise a communicator can stop its disposal process if a new handler requests to use it.

The problem is not having a different thing, let's say "thing2" wanting the very same port. This is already handled, because the port in use exception is dealt with. And this is always a user problem, since he has configured thing1 and thing2 with the same port.

What I am talking about is that the ThingHandler from thing1 is disposed and shortly afterwards initialized again. This can happen if one disabled and enables a thing, see https://github.com/openhab/openhab-core/blob/5b325aa3d210c2dfa490c676520fe755b0f3fe1c/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java#L1167-L1216

Here it doesnt matter if a communicator object or the thinghandler itself want to open the port on initialize. If it is currently in use by a thinghandler that gets disposed at the moment, then there is no way for a starting thinghandler to open the port and he will run into the port in use exception.

This applies to ALL binding which are operating on a shared resource like a port btw...

So with the change I did yesterday evening, the dispose is always called and finished before an initialize, so for now all is good within this binding.

However, one should think about the general problem of shared resources and scenarios where an initialize or dispose might take longer and one has no other choice than putting this in an external task because of this 5 second rule from openHAB.

t2000 commented:
@cpmeister Thanks for your review. Again I learned something about OSGi (reference thingy in the constructor) and that openHAB has some units on the thing-types.xmls :)

I have addressed all your comments and replied to your questions. Please have a look.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,297 @@
+        if (serialPort != null) {
+            try {
+                // send the device to sleep to preserve power and extend the lifetime of the sensor
+                sendSleep(true);
```
shouldn't you remove the event listener before you call this?

t2000 replied:
I tried this before, but somehow the underlieing library does something with the port if I de-register my listener and then my command doesn't go through.

This is just a guess from my observation. That is why I did not de-register the listener before that call here.

Also in the callback method for the listener I check for valid data, so it doesn't matter if the callback receives the reply to the sleep on shutdown.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,297 @@
+        if (logger.isDebugEnabled()) {
+            logger.debug("Will send command: {} ({})", HexUtils.bytesToHex(commandData), Arrays.toString(commandData));
+        }
+        outputStream.write(commandData, 0, commandData.length);
```
```suggestion
        outputStream.write(commandData);
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,297 @@
+        if (logger.isDebugEnabled()) {
+            logger.debug("Requesting sensor data, will send: {}", HexUtils.bytesToHex(data));
+        }
+        outputStream.write(data, 0, data.length);
```
```suggestion
        outputStream.write(data);
        outputStream.flush();
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/messages/ModeReply.java:
```diff
@@ -0,0 +1,63 @@
+@NonNullByDefault
+public class ModeReply extends SensorReply {
+
+    private byte actionType;
+    private WorkMode mode;
```
```suggestion
    private final byte actionType;
    private final WorkMode mode;
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/messages/SleepReply.java:
```diff
@@ -0,0 +1,58 @@
+@NonNullByDefault
+public class SleepReply extends SensorReply {
+
+    private byte actionType;
+    private byte sleep;
```
```suggestion
    private final byte actionType;
    private final byte sleep;
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/messages/SensorFirmwareReply.java:
```diff
@@ -0,0 +1,51 @@
+@NonNullByDefault
+public class SensorFirmwareReply extends SensorReply {
+
+    private byte year;
+    private byte month;
+    private byte day;
```
```suggestion
    private final byte year;
    private final byte month;
    private final byte day;
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/messages/WorkingPeriodReply.java:
```diff
@@ -0,0 +1,58 @@
+@NonNullByDefault
+public class WorkingPeriodReply extends SensorReply {
+
+    private byte actionType;
+    private byte period;
```
```suggestion
    private final byte actionType;
    private final byte period;
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,258 @@
+    private final Logger logger = LoggerFactory.getLogger(SDS011Handler.class);
+    private final SerialPortManager serialPortManager;
+
+    private @NonNullByDefault({}) NovaFineDustConfiguration config;
```
Either make this nullable or initialize it to some default configuration.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,258 @@
+    private final SerialPortManager serialPortManager;
+
+    private @NonNullByDefault({}) NovaFineDustConfiguration config;
+    private @NonNullByDefault({}) SDS011Communicator communicator;
```
Please make this nullable. I've had times in my own bindings that this ended up null unexpectedly.

cpmeister commented:
Please address these build warnings. Most of these can be addressed by caching the nullable field to a local variable and performing your logic on the local variable instead.
```
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[104,4308] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[112,4578] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[113,4659] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[117,4916] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[130,5598] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[154,6731] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[159,7026] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[176,7700] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[176,7707] Redundant null check: comparing '@NonNull String' against null
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[176,7723] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[180,7911] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[181,7947] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[181,7979] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[187,8268] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[187,8298] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[199,8680] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[204,8800] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:[209,8929] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[118,5004] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[119,5045] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[222,9131] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[223,9165] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[230,9356] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[231,9411] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[235,9629] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[287,11589] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:[293,11767] Potential null pointer access: this expression has a '@Nullable' type
```

t2000 commented:
I saw these warnings, but to be honest, do you really want me to write ugly code just to make this stupid analysis happy?

So this:

```
OutputStream localOutputStream = outputStream;
        if (localOutputStream != null) {
            localOutputStream.write(commandData);
            localOutputStream.flush();
        }
```

instead of

```
outputStream.write(commandData);
outputStream.flush();
```

In cases where I know that the member variable `outputStream` has definitely been checked against `null` before`?

Sure I can do that, but does this make the code more readable? Surely not.

I think the null analysis should help the developer to write better code, but it shouldn't force him to write ugly unreadable code.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,266 @@
+    private final Logger logger = LoggerFactory.getLogger(SDS011Handler.class);
+    private final SerialPortManager serialPortManager;
+
+    private @Nullable NovaFineDustConfiguration config;
```
To avoid null issues with the configs, you can change your config to be non-null and just populate the config with an instance that contains default values:
```suggestion
    private NovaFineDustConfiguration config = new NovaFineDustConfiguration();
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,298 @@
+
+    private SerialPortIdentifier portId;
+    private SDS011Handler thingHandler;
+    private @NonNullByDefault({}) SerialPort serialPort;
```
Please don't use `@NonNullByDefault({})` if possible.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/sds011protocol/SDS011Communicator.java:
```diff
@@ -0,0 +1,298 @@
+    private SDS011Handler thingHandler;
+    private @NonNullByDefault({}) SerialPort serialPort;
+
+    private @Nullable OutputStream outputStream;
+    private @Nullable InputStream inputStream;
```
If you want another way to get around the null checks. You could make these fields get populated from the constructor so you could make them non-null and final.
You would then turn your `initialize` method into a static factory method for `SDS011Communicator` instances.

t2000 replied:
I'd rather structure my code in a way that makes it look clean, i.e. in this case having the serial port and the streams within one object, rather than splitting this throughout the code just to make an analysis tool happy.
I have made the null checks now...

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,266 @@
+
+    @Override
+    public void dispose() {
+        if (pollingJob != null) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
```
Another example here of how to easily deal with the null warnings.
```suggestion
        Future<?> pollingJob = this.pollingJob;
        if (pollingJob != null) {
            pollingJob.cancel(true);
            this.pollingJob = null;
        }
```

t2000 replied:
My IDE wasn't even showing this warning.

So to me it looks like the IDE checks and the build checks are not in sync. There was even a build failure in travis (a couple of commits back) which was fine in eclipse.

Also If you insists of dealing with these warnings, I would recommend to make them errors, not warnings, so the build will fail direclty.

One side note:

If you implement is as you suggested, you will hide the class variable `pollingJob` with the local variable with the same name and this is a bad code style which can lead to errors. Even Sonoarlint marks this as something that should be changed.

t2000 commented:
Thanks for the review. I have addressed your changes.

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,270 @@
+
+        if (config.reporting) {
+            timeBetweenDataShouldArrive = Duration.ofMinutes(config.reportingInterval);
+            scheduler.schedule(() -> initializeCommunicator(WorkMode.REPORTING, timeBetweenDataShouldArrive), 0,
+                    TimeUnit.SECONDS);
```
```suggestion
            scheduler.submit(() -> initializeCommunicator(WorkMode.REPORTING, timeBetweenDataShouldArrive));
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,270 @@
+                    TimeUnit.SECONDS);
+        } else {
+            timeBetweenDataShouldArrive = Duration.ofSeconds(config.pollingInterval);
+            scheduler.schedule(() -> initializeCommunicator(WorkMode.POLLING, timeBetweenDataShouldArrive), 0,
+                    TimeUnit.SECONDS);
```
```suggestion
            scheduler.submit(() -> initializeCommunicator(WorkMode.POLLING, timeBetweenDataShouldArrive));
```

cpmeister commented on bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java:
```diff
@@ -0,0 +1,270 @@
+                connectionMonitorStartDelay.getSeconds(), timeBetweenDataShouldArrive.getSeconds(), TimeUnit.SECONDS);
+    }
+
+    private void initializeCommunicator(WorkMode mode, Duration interval) {
```
Just to be safe, I would make sure that `initialize`, `dispose`, and `initializeCommunicator` are synchronized.

t2000 replied:
How should I do that? This won't work if I write `synchronized` on those methods.

The reason is that this will synchronize of the ´ThingHandler` object and this is never the same is a `dispose` and an `initialize` are called.  The `ThingManager` will always create a new `ThingHandler` instance before calling `initialize`, so the call to `dispose` will happen on a different (old) object.

This is what I wanted to state earlier: The openHAB framework currently does NOT ensure that `dispose` and `initialize` run into each other, if one is forced to schedule a task (that runs in a different thread) from `initialize` if one needs more than 5 seconds.

So either: I do NOT schedule this task and risk the warning "ThingHandler initialize takes longer than 5 seconds", or I take the (in my case pretty unlikely) risk that `dispose` and `initialize` collide eventually.

So which option should I pick?

cpmeister replied:
@J-N-K Could you please weigh in here? I can't tell if I'm being overly concerned or not with regards to thread safety here.

kaikreuzer replied:
I can chime in here. We never have `synchronized` keywords on such methods. As @t2000 mentions, the work is done in a separate thread and it is actually intended that the framework is able to dispose the handler, while such jobs might still be running. They should NOT block a call of `dispose`, so synchronizing it would be counter-productive.

cpmeister replied:
@kaikreuzer I'm just concerned that `initializeCommunicator` would be called after `dispose` since `initializeCommunicator` is called asynchronously.

kaikreuzer replied:
Sounds pretty unlikely and even if it does, the code should be written in a way that it shouldn't harm much.
If dispose would set the [communicator to null](https://github.com/openhab/openhab-addons/pull/7528/files/e685099b38e6dda40cf3326a1182d394983790ac#diff-ab4139e8f28a3d36d178fd52c9b9404bR213) and initializeCommunicator would [not try to set the Thing status](https://github.com/openhab/openhab-addons/pull/7528/files/e685099b38e6dda40cf3326a1182d394983790ac#diff-ab4139e8f28a3d36d178fd52c9b9404bR129-R133) upon a null-communicator, it should almost be safe ;-)

t2000 commented:
Thank you for the reviews and the approvals. is there anything else that needs to be done from my side before merging this? If so, please ping me.

fwolter commented:
No, you did fine. All you have to do is to wait for another maintainer to approve your PR. That's why the CRE label has been added.

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/README.md:
```diff
@@ -0,0 +1,78 @@
+# NovaFineDust Binding
+
+This binding is for the fine dust sensor (PM Sensor) from Nova Fitness. Currently only one model is supported, the SDS011.
```
Please add a newline after every sentence throughout the README.
```suggestion
This binding is for the fine dust sensor (PM Sensor) from Nova Fitness.
Currently only one model is supported, the SDS011.
```

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## Supported Things
+
+There is only one Thing for this binding is `SDS011`.
```
```suggestion
There is only one Thing type for this binding, which is `SDS011`.
```

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## Discovery
+
+There is no automatic discovery. The Thing has to be added manually via the `.things` configuration file or via (Paper) UI.
```
```suggestion
There is no automatic discovery.
```

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/README.md:
```diff
@@ -0,0 +1,78 @@
+| parameter name    | mandatory | description                                                                           |
+|-------------------|-----------|---------------------------------------------------------------------------------------|
+| port              | yes       | the port the sensor is connected to, i.e. /dev/ttyUSB0.                              |
+| reporting         | yes       | whether the reporting mode (value=true) or polling mode should be used.               |
```
Above you say that "reporting" is the default mode. If there's a default, the parameter is not mandatory, but optional.

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## Limitations
+
+In theory once can have multiple sensors connected and distinguish them via their device ID. However, this is currently not implemented and the binding always configures any device and accepts data reportings from any device too.
```
You mean there could be multiple sensors on the same port, right?
I very much hope that it is possible to have multiple `SDS011` things that use different ports.
Maybe you can clarify this in the statement here a bit.

kaikreuzer replied:
```suggestion
In theory one can have multiple sensors connected and distinguish them via their device ID. However, this is currently not implemented and the binding always configures any device and accepts data reportings from any device too.
```

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+			</parameter>
+			<parameter name="reporting" type="boolean">
+				<default>true</default>
+				<label>Reporting If True, Otherwise Polling</label>
```
```suggestion
				<label>Mode</label>
				<options>
					<option value="true">Reporting</option>
					<option value="false">Polling</option>
				</options>
```

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+			<parameter name="reportingInterval" type="integer" min="0" max="30" unit="min">
+				<default>1</default>
+				<advanced>true</advanced>
+				<label>Time In Minutes When A Value Will Be Reported</label>
```
```suggestion
				<label>Reporting Interval</label>
```

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+			<parameter name="pollingInterval" type="integer" min="3" max="3600" unit="s">
+				<default>10</default>
+				<advanced>true</advanced>
+				<label>Time In Seconds When A Value Will Be Polled From The Device (Min 3 Seconds)</label>
```
```suggestion
				<label>Polling Interval</label>
```

kaikreuzer commented on bundles/org.openhab.binding.novafinedust/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,57 @@
+
+	<channel-type id="pm10-type">
+		<item-type>Number:Density</item-type>
+		<label>PM10</label>
```
```suggestion
		<label>PM 10</label>
```

t2000 commented:
Thanks, for the final review @kaikreuzer.
I have added 2 more commits, one with the suggestions that you posted on github and one with the clarifications that you requested on the README.

Please have a look if its OK or I missed something. Thank you!
